### ISSUE TITLE
Conditional Mapping Based on Source Value

### ISSUE THREAD

**Issue by mikeobrien** [AutoMapper/AutoMapper#1723]
We have input models for our API that use the concept of optional values, these are essentially a discriminated union that allow us to know if the value was actually passed in the json or omitted.  We then conditionally map based on those in 4.x a la:

```
.ForAllMembers(x => x
    .Condition(y => !y.SourceType.IsOptional() ||
        ((IOptional)y.SourceValue).HasValue));
```

With 5.x I can't find any way to do this anymore as I can't get ahold of the source value in a conditional. We have value mappings/converters that convert `Optional<T>` to `T`. Previously `Condition` was passed a `ResolutionContext` with the source value and type _before_ it was mapped/converted, now `ResolutionContext` has been stripped of almost everything and signature of `Condition` has `TMember` for both the source and destination value so assuming the "source" value is actually the mapped/converted value _not_ the original source (AFIACT anyways). 

Any thoughts on how we can conditionally map source values like this in 5.x?

Thx!

**Comment by lbargaoanu** [comment]
Duplicate of  #1703.

**Comment by mikeobrien** [comment]
TL;DR I'm not using nullable types to indicate missing values. I'm using a struct that wraps a value (which could be nullable) and that struct is never nullable in our models.

So this is a bit different, `Optional<T>` is a struct but it is not nullable in any of our models, so should never be null (if its the *original* source value anyways). It would be no different than if the source value was an `int` (*not* nullable) but a null was passed in as the "source" value to `Conditional()`, should never be null, should always be an `int`. AFAICT the only way it could end up being a null "source" value (Passed into `Conditional()`) is if its been mapped/converted (And the value that `Optional<T>` wrapped was null) *before* it's passed into `Conditional()` OR it isn't actually being passed into `Conditional()` at all. Either way, I need the *original* source value in order to decide what fields get mapped and what don't. This worked in 4.x but is not working the same in 5.x.

**Comment by lbargaoanu** [comment]
Sounds the same to me :) (just replace Nullable with Optional) Anyway, the idea is that now you get as source what you resolved and you should change your conditions accordingly. I'm not sure whether @jbogard would want to change this. It would surely break some other people and it makes sense the way it is now. Base the condition on what would really get assigned, not on the initial value, which is water under the bridge by now.

**Comment by mikeobrien** [comment]
> it makes sense the way it is now

Like I said, I totally get that the source value is in its final form, no arguments there. So I'm not talking about changing that. What I'm talking about is exposing the **original** source value **somewhere**, as was the case in 4.x, so that conditionals could use this to decide whether or not to map.

> Base the condition on what would really get assigned, not on the initial value, which is water under the bridge by now.

Just telling me to base it on the final source value does **not** solve my problem and does **not** fit this use case. ;) It's not "water under the bridge" because I can't conditionally map given the final source value, that information is in the **original** source value (as it should be). The use case is also very valid, not doing anything esoteric here, pretty common in an API to have optional values and therefore need to be able to conditionally map values that were not passed in.

I'm totally willing to submit a PR to expose the original source value if you are willing to accept one. And again, I'm **NOT** proposing we break `Condition()` by changing the source value from the final value to the original value, **just** exposing the original source value **somewhere** like it was before. This was originally in `ResolutionContext` is there any reason this can't be added back?

**Comment by lbargaoanu** [comment]
Why can't you? When the condition is false, the destination has a default value, right?

**Comment by lbargaoanu** [comment]
Anyway, maybe you should try that PR. Just don't touch the ResolutionContext :)

**Comment by mikeobrien** [comment]
Unfortunately you cannot always base the conditional off of a null or default value because both the default value and a null could be valid, I cannot decide whether or not the value was passed into the API based on if it's null or the default value (at least in our domain anyways, you might be able to get away with it in trivial API). Here is an example:

```csharp
public EmployeeModel
{
    public decimal? Salary { get; set; }
    public decimal? Rate { get; set; }

    // OR

    public decimal Salary { get; set; }
    public decimal Rate { get; set; }
}
```

In both these cases null and zero (the default value) could both be valid values for `Salary` and `Rate`. Basing the conditional logic off null or the default value does not allow me to determine if the value was passed in and therefore map it appropriately.

On the other hand, if we use an [option type](https://en.wikipedia.org/wiki/Option_type), you **can** determine if it was passed in and accept nulls and default values as valid values. 

```csharp
public struct Optional<T> : IOptional
{
    private T _value;

    public bool HasValue { get; private set; }

    public T Value
    {
        get { return _value; }
        set
        {
            HasValue = true;
            _value = value;
        }
    }
}

public EmployeeModel
{
    public Optional<decimal?> Salary { get; set; }
    public Optional<decimal?> Rate { get; set; }

    // OR

    public Optional<decimal> Salary { get; set; }
    public Optional<decimal> Rate { get; set; }
}

.ForAllMembers(x => x
    .Condition(y => !y.SourceType.IsOptional() ||
        ((IOptional)y.SourceValue).HasValue));
```

The mapping/converter converts the `Optional<T>` to `T` so (as it stands in 5.x) the information that would let me decide if it was passed or not is no longer available to `Condition()`.

**Comment by lbargaoanu** [comment]
```Optional<Nullable<decimal>>``` :)

**Comment by mikeobrien** [comment]
What would be an acceptable way to expose that (if `ResolutionContext` isn't the right place)? #1563 indicated that `DestinationType` was slated to be exposed in `IMemberConfiguration`, would this be an appropriate place?

**Comment by mikeobrien** [comment]
Yeah, optional and nullable are definitely not the same thing. I know people try to use them that way but they are semantically different.

**Comment by lbargaoanu** [comment]
I'm afraid it would have to be another Condition overload.

**Comment by mikeobrien** [comment]
Ah ok, you guys ok with that? My only concern is that there could end up being a million overloads for things people want to pass, so I would understand if you wanted to limit that. But if you're ok with it I'll go that route.

**Comment by lbargaoanu** [comment]
It's not pretty :) It's @jbogard's decision.

**Comment by mikeobrien** [comment]
Ok cool, I'll wait to hear back from him then.

**Comment by mikeobrien** [comment]
@jbogard would you be open to a PR to enable this?

**Comment by jbogard** [comment]
Yeah, the trick is understanding where the conditional is evaluated. But I'm open.
